futures-util/src/task/atomic_waker.rs

```diff
@@ -198,80 +238,91 @@ impl AtomicWaker {
+                        debug_assert_eq!(actual, REGISTERING | WAKING);
+
+                        // Take the waker to wake once the atomic operation has
+                        // completed.
+                        let waker = (*self.waker.get()).take().unwrap();
+
+                        // Just swap, because no one could change state while state == `REGISTERING` | `WAKING`.
+                        self.state.swap(EMPTY, AcqRel);
+
+                        // The atomic swap was complete, now
+                        // wake the task and return.
+                        waker.wake();
                     }
                 }
             }
-            WAKING => {
-                // Currently in the process of waking the task, i.e.,
-                // `wake` is currently being called on the old task handle.
-                // So, we call wake on the new waker
-                lw.wake();
-            }
-            state => {
-                // In this case, a concurrent thread is holding the
-                // "registering" lock. This probably indicates a bug in the
-                // caller's code as racing to call `register` doesn't make much
-                // sense.
-                //
-                // We just want to maintain memory safety. It is ok to drop the
-                // call to `register`.
-                debug_assert!(
-                    state == REGISTERING ||
-                    state == REGISTERING | WAKING);
-            }
+        }
+
+        if state == WAKING || state == WAKING | HOLDS {
+            // Currently in the process of waking the task, i.e.,
+            // `wake` is currently being called on the old task handle.
+            // So, we call wake on the new waker
+            lw.wake();
+        } else {
+            // In this case, a concurrent thread is holding the
+            // "registering" lock. This probably indicates a bug in the
+            // caller's code as racing to call `register` doesn't make much
+            // sense.
+            //
+            // We just want to maintain memory safety. It is ok to drop the
+            // call to `register`.
+            debug_assert!(
+                state == REGISTERING ||
+                state == REGISTERING | WAKING);
         }
     }
 
     /// Calls `wake` on the last `Waker` passed to `register`.
     ///
     /// If `register` has not been called yet, then this does nothing.
     pub fn wake(&self) {
+        let state = self.state.load(Acquire);
+
+        if state == EMPTY || state & WAKING != 0 {
+            fence(Release);
+            return;
+        }
+
         // AcqRel ordering is used in order to acquire the value of the `task`
         // cell as well as to establish a `release` ordering with whatever
         // memory the `AtomicWaker` is associated with.
         match self.state.fetch_or(WAKING, AcqRel) {
-            WAITING => {
+            state if state == EMPTY || state == HOLDS => {
```

Comment:
nit: `EMPTY | HOLDS => { ... }`

Reply:
I need to update `state` variable here. I'll update code to make it more clear.